File: frontend/src/components/benchmark/content/radix_sort/implementations/radix_sort.php
<?php

class RadixSort {
    private function countingSort(array &$arr, int $exp): void {
        $n = count($arr);
        $output = array_fill(0, $n, 0);
        $count = array_fill(0, 10, 0);

        for ($i = 0; $i < $n; $i++) {
            $index = intdiv($arr[$i], $exp) % 10;
            $count[$index]++;
        }

        for ($i = 1; $i < 10; $i++) {
            $count[$i] += $count[$i - 1];
        }

        for ($i = $n - 1; $i >= 0; $i--) {
            $index = intdiv($arr[$i], $exp) % 10;
            $output[$count[$index] - 1] = $arr[$i];
            $count[$index]--;
        }

        for ($i = 0; $i < $n; $i++) {
            $arr[$i] = $output[$i];
        }
    }

    public function sort(array &$arr): void {
        if (empty($arr)) return;

        $maxVal = max($arr);
        $exp = 1;

        while (intdiv($maxVal, $exp) > 0) {
            $this->countingSort($arr, $exp);
            $exp *= 10;
        }
    }

    public function sortNegative(array &$arr): void {
        if (empty($arr)) return;

        $negatives = [];
        $positives = [];

        foreach ($arr as $x) {
            if ($x < 0) $negatives[] = -$x;
            else $positives[] = $x;
        }

        $this->sort($negatives);
        $this->sort($positives);

        $index = 0;
        for ($i = count($negatives) - 1; $i >= 0; $i--) {
            $arr[$index++] = -$negatives[$i];
        }
        for ($i = 0; $i < count($positives); $i++) {
            $arr[$index++] = $positives[$i];
        }
    }
}

?>

File: frontend/src/components/benchmark/content/radix_sort/implementations/radix_sort.rs
pub struct RadixSort;

impl RadixSort {
    fn counting_sort(arr: &mut [i32], exp: i32) {
        let n = arr.len();
        let mut output = vec![0; n];
        let mut count = vec![0; 10];

        for i in 0..n {
            let index = ((arr[i] / exp) % 10) as usize;
            count[index] += 1;
        }

        for i in 1..10 {
            count[i] += count[i - 1];
        }

        for i in (0..n).rev() {
            let index = ((arr[i] / exp) % 10) as usize;
            output[count[index] - 1] = arr[i];
            count[index] -= 1;
        }

        arr.copy_from_slice(&output);
    }

    pub fn sort(arr: &mut [i32]) {
        if arr.is_empty() {
            return;
        }

        let max_val = *arr.iter().max().unwrap();
        let mut exp = 1;

        while max_val / exp > 0 {
            Self::counting_sort(arr, exp);
            exp *= 10;
        }
    }

    pub fn sort_negative(arr: &mut [i32]) {
        if arr.is_empty() {
            return;
        }

        let mut negatives = Vec::new();
        let mut positives = Vec::new();

        for &x in arr.iter() {
            if x < 0 {
                negatives.push(-x);
            } else {
                positives.push(x);
            }
        }

        Self::sort(&mut negatives);
        Self::sort(&mut positives);

        let mut index = 0;
        for i in (0..negatives.len()).rev() {
            arr[index] = -negatives[i];
            index += 1;
        }
        for i in 0..positives.len() {
            arr[index] = positives[i];
            index += 1;
        }
    }
}